witness/golang/cmd/witness/internal/witness/witness.go

```diff
@@ -155,75 +154,153 @@ func (w *Witness) Update(ctx context.Context, logID string, chkptRaw []byte, pro
+		default:
+			logV := logverifier.New(logInfo.Hasher)
+			if err := logV.VerifyConsistencyProof(int64(chkptOld.Size), int64(chkptNew.Size), chkptOld.Hash, chkptNew.Hash, proof); err != nil {
+				// Complain if the checkpoints aren't consistent.
+				return chkptOld.Size, fmt.Errorf("failed to verify consistency proof: %v", err)
+			}
+			// If the consistency proof is good we store chkptRaw.
+			return chkptOld.Size, w.setChkptData(tx, logID, chkptRaw, nil)
 		}
-		// If the consistency proof is good we store chkptRaw.
-		return p.Size, w.setCheckpoint(tx, logID, c)
-	case chkpt.Size == p.Size:
-		if !bytes.Equal(chkpt.Hash, p.Hash) {
-			return p.Size, fmt.Errorf("checkpoint for same size log with differing hash (got %x, have %x)", chkpt.Hash, p.Hash)
+	case chkptNew.Size == chkptOld.Size:
+		if !bytes.Equal(chkptNew.Hash, chkptOld.Hash) {
+			return chkptOld.Size, fmt.Errorf("checkpoint for same size log with differing hash (got %x, have %x)", chkptNew.Hash, chkptOld.Hash)
 		}
-		// Identical to the current latest, so "ok"
-		return p.Size, nil
+		// If it's identical to the latest do nothing.
+		return chkptOld.Size, nil
 	default:
 		panic("unreachable")
 	}
 }
 
-// getLatestChkpt returns the latest checkpoint for a given log.
-func (w *Witness) getLatestChkpt(queryRow func(query string, args ...interface{}) *sql.Row, logID string) (Chkpt, error) {
-	row := queryRow("SELECT raw, size FROM chkpts WHERE logID = ?", logID)
+// getLatestChkptData returns the raw stored data for the latest checkpoint and
+// its associated compact range, if one exists, for a given log.
+func (w *Witness) getLatestChkptData(queryRow func(query string, args ...interface{}) *sql.Row, logID string) ([]byte, []byte, error) {
+	row := queryRow("SELECT chkpt, range FROM chkpts WHERE logID = ?", logID)
 	if err := row.Err(); err != nil {
-		return Chkpt{}, err
+		return nil, nil, err
 	}
-	var maxChkpt Chkpt
-	if err := row.Scan(&maxChkpt.Raw, &maxChkpt.Size); err != nil {
+	var rawChkpt []byte
+	var rawPf []byte
+	if err := row.Scan(&rawChkpt, &rawPf); err != nil {
 		if err == sql.ErrNoRows {
-			return Chkpt{}, status.Errorf(codes.NotFound, "no checkpoint for log %q", logID)
+			return nil, nil, status.Errorf(codes.NotFound, "no checkpoint for log %q", logID)
 		}
-		return Chkpt{}, err
+		return nil, nil, err
+	}
+	return rawChkpt, rawPf, nil
+}
+
+// verifyRange verifies the new checkpoint against the stored and given compact
+// range and outputs the new compact range if verification succeeds.
+func (w *Witness) verifyRange(chkptNew *log.Checkpoint, chkptOld *log.Checkpoint, h hashers.LogHasher, curRangeRaw [][]byte, newRangeRaw [][]byte) ([][]byte, error) {
+	rf := compact.RangeFactory{Hash: h.HashChildren}
+	curRange, err := rf.NewRange(0, chkptOld.Size, curRangeRaw)
+	if err != nil {
+		return nil, fmt.Errorf("can't form current compact range: %v", err)
+	}
+	// As a sanity check, make sure the old checkpoint is consistent with
+	// the current range.
```

Comment:
What is the distinction between "old" and "current" here? Should we reduce this to only one (or zero?) adjective? E.g.

`// Make sure that compact range is consistent with the checkpoint.`

Reply:
I went with one adjective I think ("stored").